ocamlPackages tree-wide: Move buildInputs that should be nativeBuildInputs

ulrikstrid opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

To make cross compilation better we need to move a bunch of stuff to `nativeBuildInputs`.
Another good reason is that we should get much smaller closures when building ocaml projects.
This PR supersedes #145448 I think.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ulrikstrid committed:
ocamlPackages tree-wide: Move buildInputs that should be nativeBuildInputs

To keep this for the future we also strictDeps where possible, including for janePackages, topkg, oasis and ocamlbuild.
This makes some closures significantly smaller and makes cross compilation easier

symphorien commented on pkgs/development/ocaml-modules/bolt/default.nix:
```diff
@@ -18,7 +18,8 @@ stdenv.mkDerivation rec {
 
-  buildInputs = [ ocaml findlib ocamlbuild which camlp4 ];
+  nativeBuildInputs = [ ocaml findlib ocamlbuild which ];
+  buildInputs = [ camlp4 ];
```
I suspect this belongs to nativeBuildInput, as it's a preprocessor.

ulrikstrid replied:
Me and @anmonteiro discussed this and we believe `camlp4` will be built and ran while building like ppxes. I would love to be proved wrong and be able to move more things to `nativeBuildInputs`

symphorien replied:
Now that strictDeps is set, it only builds with camlp4 in nativeBuildInputs. That answers the question :)

symphorien reviewed (commented):
Thanks for the PR. How did you test it?

It would be nice to add `strictDeps = true` on every derivation which you fix. This allows to test if the partition between buildInputs and nativeBuildInputs is correct without cross compiling. This is the one true way to know if ppx libs belong to buildInputs or to nativeBuildInputs. In case strictDeps do not work at all in problematic cases (everything involving gtk3 is problematic for example), you can leave it to false, of course. But such a PR would bitrot very fast if we cannot set strictDeps to true on a large proportion of them.

ulrikstrid commented:
I've been running `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` and building LIGO statically in a separate project using this patch. It's not fully getting everything but I found a bunch of stuff that I fixed and amended.
I also did a patch to set `strictDeps` in `buildDunePackage` just to test it out, but I could add it to all the packages that I've touched in this PR.

symphorien commented:
adding strictDeps = true to buildDunePackage is a good idea if there are not too many exceptions.
Yes  it would be better to set it to all packages touched by this PR, otherwise there is no way to check that what you changed is correct, and it would bitrot rapidly.

ulrikstrid commented:
@symphorien I have now added `strictDeps = true;` to all the packages I dared setting it for (skipped the SDL and gnome packages)

ulrikstrid commented:
> adding strictDeps = true to buildDunePackage is a good idea if there are not too many exceptions. Yes it would be better to set it to all packages touched by this PR, otherwise there is no way to check that what you changed is correct, and it would bitrot rapidly.

Might be possible to add it to `buildDunePackage` after we land this PR and fix up the packages that fails at that point.

ulrikstrid commented:
Adding `strictDeps = true;` showed a lot of issues using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` but I will have to fix those tomorrow.

symphorien commented:
> > adding strictDeps = true to buildDunePackage is a good idea if there are not too many exceptions. Yes it would be better to set it to all packages touched by this PR, otherwise there is no way to check that what you changed is correct, and it would bitrot rapidly.
> 
> Might be possible to add it to `buildDunePackage` after we land this PR and fix up the packages that fails at that point.

looks good to me.

ulrikstrid commented on pkgs/development/ocaml-modules/lwt-exit/default.nix:
```diff
@@ -24,7 +24,8 @@ buildDunePackage rec {
 
-  doCheck = true;
+  # for some reason this never exits
+  doCheck = false;
```
I'm not sure if this issue is something I introduced or not but it's broken, at least on my machine

ulrikstrid commented:
Added some more fixes and now everything except `torch` (which I reverted all my changes from) builds correctly when using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`.

So I think this is ready for review and then we can decide what to do with the history, maybe squash everything into 1 commit?

ulrikstrid commented:
`nixpkgs-review` results

<details><summary>5 packages marked as broken and skipped:</summary>
ocamlPackages.accessor_async ocamlPackages.accessor_core ocamlPackages.ocaml_extlib-1-7-7 ocamlPackages.reason-native.qcheck-rely ocamlPackages.wodan-irmin
</details>

<details><summary>1 package failed to build:</summary>
ocamlPackages.torch
</details>

<details><summary>737 packages built:</summary>
acgtk alt-ergo beluga coccinelle comby coqPackages.QuickChick coqPackages.coq-elpi coqPackages.graph-theory coqPackages.hierarchy-builder coqPackages.mathcomp-analysis coqPackages.semantics coqPackages.serapi coq_8_10 coq_8_11 coq_8_12 coq_8_5 coq_8_6 coq_8_7 coq_8_8 coq_8_9 csvtool cubicle cvehound diffoscope dot-merlin-reader dune-release eff flitter flow framac fstar glsurf google-drive-ocamlfuse haxe haxePackages.hxcpp haxePackages.hxcs haxePackages.hxjava haxePackages.hxnodejs_4 haxe_4_0 haxe_4_1 hol_light ikos jackline libbap libguestfs libguestfs-with-appliance ligo liquidsoap mldonkey monotoneViz obelisk ocaml-crunch ocaml-top ocamlPackages.accessor ocamlPackages.accessor_base ocamlPackages.alcotest ocamlPackages.alcotest-lwt ocamlPackages.alcotest-mirage ocamlPackages.angstrom ocamlPackages.angstrom-async ocamlPackages.angstrom-lwt-unix ocamlPackages.angstrom-unix ocamlPackages.apron ocamlPackages.arp ocamlPackages.asn1-combinators ocamlPackages.astring ocamlPackages.async ocamlPackages.async_extra ocamlPackages.async_find ocamlPackages.async_inotify ocamlPackages.async_interactive ocamlPackages.async_js ocamlPackages.async_kernel ocamlPackages.async_rpc_kernel ocamlPackages.async_sendfile ocamlPackages.async_shell ocamlPackages.async_smtp ocamlPackages.async_ssl ocamlPackages.async_unix ocamlPackages.atd ocamlPackages.atdgen ocamlPackages.awa ocamlPackages.awa-lwt ocamlPackages.awa-mirage ocamlPackages.bap ocamlPackages.base ocamlPackages.base64 ocamlPackages.base_bigstring ocamlPackages.base_quickcheck ocamlPackages.batteries ocamlPackages.benchmark ocamlPackages.bigarray-overlap ocamlPackages.bignum ocamlPackages.bigstringaf ocamlPackages.bin_prot ocamlPackages.biniou ocamlPackages.biocaml ocamlPackages.bisect_ppx ocamlPackages.bistro ocamlPackages.bitv ocamlPackages.bls12-381-legacy ocamlPackages.bls12-381-unix ocamlPackages.bonsai ocamlPackages.bos ocamlPackages.bz2 ocamlPackages.ca-certs ocamlPackages.ca-certs-nss ocamlPackages.calendar ocamlPackages.callipyge ocamlPackages.camlimages ocamlPackages.camlpdf ocamlPackages.camlzip ocamlPackages.camomile ocamlPackages.caqti ocamlPackages.caqti-async ocamlPackages.caqti-driver-mariadb ocamlPackages.caqti-driver-postgresql ocamlPackages.caqti-driver-sqlite3 ocamlPackages.caqti-dynload ocamlPackages.caqti-lwt ocamlPackages.caqti-type-calendar ocamlPackages.carton ocamlPackages.carton-git ocamlPackages.carton-lwt ocamlPackages.cfstream ocamlPackages.chacha ocamlPackages.charInfo_width ocamlPackages.checkseum ocamlPackages.cinaps ocamlPackages.cmdliner ocamlPackages.cohttp ocamlPackages.cohttp-async ocamlPackages.cohttp-lwt ocamlPackages.cohttp-lwt-unix ocamlPackages.cohttp-mirage ocamlPackages.coin ocamlPackages.color ocamlPackages.conduit ocamlPackages.conduit-async ocamlPackages.conduit-lwt ocamlPackages.conduit-lwt-unix ocamlPackages.conduit-mirage ocamlPackages.config-file ocamlPackages.containers ocamlPackages.containers-data ocamlPackages.cooltt ocamlPackages.core ocamlPackages.core_bench ocamlPackages.core_extended ocamlPackages.core_kernel ocamlPackages.cow ocamlPackages.cpdf ocamlPackages.crowbar ocamlPackages.crunch ocamlPackages.cstruct ocamlPackages.cstruct-async ocamlPackages.cstruct-lwt ocamlPackages.cstruct-sexp ocamlPackages.cstruct-unix ocamlPackages.csvfields ocamlPackages.ctypes ocamlPackages.cudf ocamlPackages.curly ocamlPackages.dap ocamlPackages.data-encoding ocamlPackages.dbf ocamlPackages.decompress ocamlPackages.delimited_parsing ocamlPackages.diet ocamlPackages.digestif ocamlPackages.dispatch ocamlPackages.dns ocamlPackages.dns-certify ocamlPackages.dns-cli ocamlPackages.dns-client ocamlPackages.dns-mirage ocamlPackages.dns-resolver ocamlPackages.dns-server ocamlPackages.dns-stub ocamlPackages.dns-tsig ocamlPackages.dolmen ocamlPackages.dolog ocamlPackages.domain-name ocamlPackages.dose3 ocamlPackages.duff ocamlPackages.dum ocamlPackages.duration ocamlPackages.earlybird ocamlPackages.ecaml ocamlPackages.eigen ocamlPackages.elina ocamlPackages.eliom ocamlPackages.elpi ocamlPackages.email_message ocamlPackages.emile ocamlPackages.encore ocamlPackages.eqaf ocamlPackages.erm_xml ocamlPackages.erm_xmpp ocamlPackages.ethernet ocamlPackages.expect_test_helpers_async ocamlPackages.expect_test_helpers_core ocamlPackages.ezjsonm ocamlPackages.ezresto ocamlPackages.ezresto-directory ocamlPackages.ezxmlm ocamlPackages.faraday ocamlPackages.faraday-async ocamlPackages.faraday-lwt ocamlPackages.faraday-lwt-unix ocamlPackages.farfadet ocamlPackages.ff ocamlPackages.ff-pbt ocamlPackages.fiat-p256 ocamlPackages.fieldslib ocamlPackages.fileutils ocamlPackages.fmt ocamlPackages.fontconfig ocamlPackages.fpath ocamlPackages.frontc ocamlPackages.functoria ocamlPackages.functoria-runtime ocamlPackages.functory ocamlPackages.gapi_ocaml ocamlPackages.gen ocamlPackages.getopt ocamlPackages.gettext-camomile ocamlPackages.gettext-stub ocamlPackages.gg ocamlPackages.git ocamlPackages.git-cohttp ocamlPackages.git-cohttp-unix ocamlPackages.git-paf ocamlPackages.git-unix ocamlPackages.gluten ocamlPackages.gluten-lwt ocamlPackages.gluten-lwt-unix ocamlPackages.gmap ocamlPackages.gmetadom ocamlPackages.gnuplot ocamlPackages.graphql ocamlPackages.graphql-cohttp ocamlPackages.graphql-lwt ocamlPackages.graphql_parser ocamlPackages.graphql_ppx ocamlPackages.h2 ocamlPackages.hack_parallel ocamlPackages.hacl-star ocamlPackages.hacl-star-raw ocamlPackages.hacl_x25519 ocamlPackages.hex ocamlPackages.higher_kinded ocamlPackages.higlo ocamlPackages.hkdf ocamlPackages.hmap ocamlPackages.hpack ocamlPackages.httpaf ocamlPackages.hxd ocamlPackages.imagelib ocamlPackages.incr_dom ocamlPackages.incr_map ocamlPackages.incr_select ocamlPackages.incremental ocamlPackages.index ocamlPackages.inifiles ocamlPackages.inotify ocamlPackages.io-page ocamlPackages.io-page-unix ocamlPackages.ipaddr ocamlPackages.ipaddr-cstruct ocamlPackages.ipaddr-sexp ocamlPackages.iri ocamlPackages.irmin ocamlPackages.irmin-chunk ocamlPackages.irmin-containers ocamlPackages.irmin-fs ocamlPackages.irmin-git ocamlPackages.irmin-graphql ocamlPackages.irmin-http ocamlPackages.irmin-layers ocamlPackages.irmin-mirage ocamlPackages.irmin-mirage-git ocamlPackages.irmin-mirage-graphql ocamlPackages.irmin-pack ocamlPackages.irmin-test ocamlPackages.irmin-unix ocamlPackages.irmin-watcher ocamlPackages.iso8601 ocamlPackages.iter ocamlPackages.javalib ocamlPackages.jingoo ocamlPackages.js_of_ocaml ocamlPackages.js_of_ocaml-compiler ocamlPackages.js_of_ocaml-lwt ocamlPackages.js_of_ocaml-ppx ocamlPackages.js_of_ocaml-ppx_deriving_json ocamlPackages.js_of_ocaml-tyxml ocamlPackages.json-data-encoding ocamlPackages.json-data-encoding-bson ocamlPackages.jsonm ocamlPackages.jst-config ocamlPackages.junit ocamlPackages.junit_alcotest ocamlPackages.junit_ounit ocamlPackages.jwto ocamlPackages.kafka_lwt ocamlPackages.ke ocamlPackages.lablgl ocamlPackages.lablgtk ocamlPackages.lablgtk-extras ocamlPackages.labltk ocamlPackages.lambdaTerm ocamlPackages.lambdasoup ocamlPackages.lens ocamlPackages.letsencrypt ocamlPackages.letsencrypt-app ocamlPackages.letsencrypt-dns ocamlPackages.llvm ocamlPackages.logs ocamlPackages.lru ocamlPackages.lsp ocamlPackages.lua-ml ocamlPackages.lustre-v6 ocamlPackages.lutils ocamlPackages.luv ocamlPackages.lwt-exit ocamlPackages.lwt_ppx ocamlPackages.lwt_react ocamlPackages.macaddr ocamlPackages.macaddr-cstruct ocamlPackages.macaddr-sexp ocamlPackages.macaque ocamlPackages.mariadb ocamlPackages.markup ocamlPackages.mccs ocamlPackages.mdx ocamlPackages.merlin ocamlPackages.metrics ocamlPackages.metrics-influx ocamlPackages.metrics-lwt ocamlPackages.metrics-mirage ocamlPackages.metrics-unix ocamlPackages.mew_vi ocamlPackages.mimic ocamlPackages.mirage ocamlPackages.mirage-block ocamlPackages.mirage-block-combinators ocamlPackages.mirage-block-ramdisk ocamlPackages.mirage-block-unix ocamlPackages.mirage-bootvar-unix ocamlPackages.mirage-bootvar-xen ocamlPackages.mirage-channel ocamlPackages.mirage-console ocamlPackages.mirage-console-unix ocamlPackages.mirage-crypto ocamlPackages.mirage-crypto-ec ocamlPackages.mirage-crypto-pk ocamlPackages.mirage-crypto-rng ocamlPackages.mirage-crypto-rng-async ocamlPackages.mirage-crypto-rng-mirage ocamlPackages.mirage-device ocamlPackages.mirage-flow ocamlPackages.mirage-flow-combinators ocamlPackages.mirage-flow-unix ocamlPackages.mirage-fs ocamlPackages.mirage-kv ocamlPackages.mirage-logs ocamlPackages.mirage-nat ocamlPackages.mirage-net ocamlPackages.mirage-net-xen ocamlPackages.mirage-profile ocamlPackages.mirage-protocols ocamlPackages.mirage-random ocamlPackages.mirage-random-test ocamlPackages.mirage-runtime ocamlPackages.mirage-stack ocamlPackages.mirage-time-unix ocamlPackages.mirage-types ocamlPackages.mirage-types-lwt ocamlPackages.mirage-unix ocamlPackages.mirage-vnetif ocamlPackages.mirage-xen ocamlPackages.mlgmpidl ocamlPackages.mrmime ocamlPackages.mtime ocamlPackages.mustache ocamlPackages.netchannel ocamlPackages.nocrypto ocamlPackages.noise ocamlPackages.notty ocamlPackages.npy ocamlPackages.num ocamlPackages.ocaml-lsp ocamlPackages.ocaml-monadic ocamlPackages.ocaml-print-intf ocamlPackages.ocaml-r ocamlPackages.ocaml-sat-solvers ocamlPackages.ocaml_expat ocamlPackages.ocaml_extlib ocamlPackages.ocaml_gettext ocamlPackages.ocaml_libvirt ocamlPackages.ocaml_mysql ocamlPackages.ocaml_oasis ocamlPackages.ocamlgraph_gtk ocamlPackages.ocamlmod ocamlPackages.ocamlnet ocamlPackages.ocb-stubblr ocamlPackages.ocf ocamlPackages.ocf_ppx ocamlPackages.ocp-build ocamlPackages.ocp-indent ocamlPackages.ocp-index ocamlPackages.ocp-ocamlres ocamlPackages.ocplib-json-typed ocamlPackages.ocplib-json-typed-browser ocamlPackages.ocplib-json-typed-bson ocamlPackages.ocplib-simplex ocamlPackages.ocsigen-start ocamlPackages.ocsigen-toolkit ocamlPackages.ocsigen_deriving ocamlPackages.ocsigen_server ocamlPackages.ocurl ocamlPackages.odoc ocamlPackages.odoc-parser ocamlPackages.omd ocamlPackages.opam-core ocamlPackages.opam-format ocamlPackages.opam-repository ocamlPackages.opam-state ocamlPackages.opium ocamlPackages.opium_kernel ocamlPackages.otfm ocamlPackages.otoml ocamlPackages.otr ocamlPackages.ounit ocamlPackages.owl ocamlPackages.paf ocamlPackages.paf-cohttp ocamlPackages.paf-le ocamlPackages.parany ocamlPackages.parse-argv ocamlPackages.parsexp ocamlPackages.patience_diff ocamlPackages.pbkdf ocamlPackages.pcap-format ocamlPackages.pecu ocamlPackages.pgocaml ocamlPackages.pgocaml_ppx ocamlPackages.pgsolver ocamlPackages.phylogenetics ocamlPackages.piaf ocamlPackages.piqi ocamlPackages.piqi-ocaml ocamlPackages.posix-base ocamlPackages.posix-socket ocamlPackages.posix-types ocamlPackages.posixat ocamlPackages.pp ocamlPackages.ppx_accessor ocamlPackages.ppx_assert ocamlPackages.ppx_bap ocamlPackages.ppx_base ocamlPackages.ppx_bench ocamlPackages.ppx_bin_prot ocamlPackages.ppx_bitstring ocamlPackages.ppx_blob ocamlPackages.ppx_cold ocamlPackages.ppx_compare ocamlPackages.ppx_cstruct ocamlPackages.ppx_cstubs ocamlPackages.ppx_custom_printf ocamlPackages.ppx_deriving ocamlPackages.ppx_deriving_cmdliner ocamlPackages.ppx_deriving_protobuf ocamlPackages.ppx_deriving_rpc ocamlPackages.ppx_deriving_yaml ocamlPackages.ppx_deriving_yojson ocamlPackages.ppx_enumerate ocamlPackages.ppx_expect ocamlPackages.ppx_fields_conv ocamlPackages.ppx_fixed_literal ocamlPackages.ppx_gen_rec ocamlPackages.ppx_hash ocamlPackages.ppx_here ocamlPackages.ppx_import ocamlPackages.ppx_inline_test ocamlPackages.ppx_irmin ocamlPackages.ppx_jane ocamlPackages.ppx_js_style ocamlPackages.ppx_let ocamlPackages.ppx_log ocamlPackages.ppx_module_timer ocamlPackages.ppx_optcomp ocamlPackages.ppx_optional ocamlPackages.ppx_pattern_bind ocamlPackages.ppx_pipebang ocamlPackages.ppx_python ocamlPackages.ppx_repr ocamlPackages.ppx_sexp_conv ocamlPackages.ppx_sexp_message ocamlPackages.ppx_sexp_value ocamlPackages.ppx_stable ocamlPackages.ppx_string ocamlPackages.ppx_tools ocamlPackages.ppx_typerep_conv ocamlPackages.ppx_variants_conv ocamlPackages.ppx_yojson_conv_lib ocamlPackages.ppxlib ocamlPackages.prettym ocamlPackages.printbox ocamlPackages.printbox-text ocamlPackages.progress ocamlPackages.protocol_version_header ocamlPackages.psq ocamlPackages.ptime ocamlPackages.pyml ocamlPackages.pythonlib ocamlPackages.qcheck ocamlPackages.qcheck-alcotest ocamlPackages.qcheck-ounit ocamlPackages.qtest ocamlPackages.randomconv ocamlPackages.rdbg ocamlPackages.re ocamlPackages.re2 ocamlPackages.re2_stable ocamlPackages.react ocamlPackages.reactivedata ocamlPackages.reason ocamlPackages.reason-native.console ocamlPackages.reason-native.dir ocamlPackages.reason-native.file-context-printer ocamlPackages.reason-native.fp ocamlPackages.reason-native.pastel ocamlPackages.reason-native.pastel-console ocamlPackages.reason-native.refmterr ocamlPackages.reason-native.rely ocamlPackages.reason-native.rely-junit-reporter ocamlPackages.repr ocamlPackages.resource_cache ocamlPackages.resto ocamlPackages.resto-acl ocamlPackages.resto-cohttp ocamlPackages.resto-cohttp-client ocamlPackages.resto-cohttp-self-serving-client ocamlPackages.resto-cohttp-server ocamlPackages.resto-directory ocamlPackages.resto-json ocamlPackages.rfc7748 ocamlPackages.rope ocamlPackages.rosetta ocamlPackages.rpclib ocamlPackages.rpclib-lwt ocamlPackages.rresult ocamlPackages.sawja ocamlPackages.secp256k1 ocamlPackages.secp256k1-internal ocamlPackages.sedlex ocamlPackages.sedlex_2 ocamlPackages.sexp ocamlPackages.sexp_diff_kernel ocamlPackages.sexp_macro ocamlPackages.sexp_pretty ocamlPackages.sexp_select ocamlPackages.sexplib ocamlPackages.sha ocamlPackages.shared-memory-ring ocamlPackages.shared-memory-ring-lwt ocamlPackages.shell ocamlPackages.shexp ocamlPackages.sodium ocamlPackages.sosa ocamlPackages.spawn ocamlPackages.splay_tree ocamlPackages.splittable_random ocamlPackages.srt ocamlPackages.stdcompat ocamlPackages.stdint ocamlPackages.stdio ocamlPackages.stog ocamlPackages.stringext ocamlPackages.syslog-message ocamlPackages.tar ocamlPackages.tar-unix ocamlPackages.tcpip ocamlPackages.tcslib ocamlPackages.terminal ocamlPackages.terminal_size ocamlPackages.textutils ocamlPackages.tezos-010-PtGRANAD-test-helpers ocamlPackages.tezos-011-PtHangz2-test-helpers ocamlPackages.tezos-base ocamlPackages.tezos-base-test-helpers ocamlPackages.tezos-clic ocamlPackages.tezos-client-010-PtGRANAD ocamlPackages.tezos-client-011-PtHangz2 ocamlPackages.tezos-client-base ocamlPackages.tezos-context ocamlPackages.tezos-crypto ocamlPackages.tezos-embedded-protocol-010-PtGRANAD ocamlPackages.tezos-embedded-protocol-demo-noops ocamlPackages.tezos-error-monad ocamlPackages.tezos-event-logging ocamlPackages.tezos-event-logging-test-helpers ocamlPackages.tezos-hacl-glue-unix ocamlPackages.tezos-legacy-store ocamlPackages.tezos-lmdb ocamlPackages.tezos-lwt-result-stdlib ocamlPackages.tezos-micheline ocamlPackages.tezos-mockup-proxy ocamlPackages.tezos-mockup-registration ocamlPackages.tezos-p2p ocamlPackages.tezos-p2p-services ocamlPackages.tezos-protocol-010-PtGRANAD ocamlPackages.tezos-protocol-010-PtGRANAD-parameters ocamlPackages.tezos-protocol-011-PtHangz2 ocamlPackages.tezos-protocol-011-PtHangz2-parameters ocamlPackages.tezos-protocol-alpha ocamlPackages.tezos-protocol-compiler ocamlPackages.tezos-protocol-demo-noops ocamlPackages.tezos-protocol-environment ocamlPackages.tezos-protocol-environment-sigs ocamlPackages.tezos-protocol-environment-structs ocamlPackages.tezos-protocol-plugin-010-PtGRANAD ocamlPackages.tezos-protocol-plugin-011-PtHangz2 ocamlPackages.tezos-protocol-plugin-alpha ocamlPackages.tezos-protocol-updater ocamlPackages.tezos-proxy ocamlPackages.tezos-requester ocamlPackages.tezos-rpc ocamlPackages.tezos-rpc-http ocamlPackages.tezos-rpc-http-client ocamlPackages.tezos-rpc-http-client-unix ocamlPackages.tezos-sapling ocamlPackages.tezos-shell ocamlPackages.tezos-shell-context ocamlPackages.tezos-shell-services ocamlPackages.tezos-shell-services-test-helpers ocamlPackages.tezos-signer-backends ocamlPackages.tezos-signer-services ocamlPackages.tezos-stdlib ocamlPackages.tezos-stdlib-unix ocamlPackages.tezos-store ocamlPackages.tezos-test-helpers ocamlPackages.tezos-validation ocamlPackages.tezos-version ocamlPackages.tezos-workers ocamlPackages.time_now ocamlPackages.timezone ocamlPackages.tls ocamlPackages.tls-async ocamlPackages.tls-mirage ocamlPackages.toml ocamlPackages.topological_sort ocamlPackages.tsdl ocamlPackages.tsort ocamlPackages.tuntap ocamlPackages.twt ocamlPackages.typerep ocamlPackages.tyxml ocamlPackages.uchar ocamlPackages.uecc ocamlPackages.ulex ocamlPackages.unstrctrd ocamlPackages.uri ocamlPackages.uri-sexp ocamlPackages.utop ocamlPackages.uucd ocamlPackages.uucp ocamlPackages.uuidm ocamlPackages.uunf ocamlPackages.uuseg ocamlPackages.uutf ocamlPackages.uuuu ocamlPackages.variantslib ocamlPackages.vcaml ocamlPackages.vchan ocamlPackages.vg ocamlPackages.virtual_dom ocamlPackages.visitors ocamlPackages.wayland ocamlPackages.webbrowser ocamlPackages.webmachine ocamlPackages.wodan ocamlPackages.wodan-unix ocamlPackages.x509 ocamlPackages.xenstore ocamlPackages.xenstore-tool ocamlPackages.xenstore_transport ocamlPackages.xml-light ocamlPackages.xmlm ocamlPackages.xtmpl ocamlPackages.xtmpl_ppx ocamlPackages.yaml ocamlPackages.yaml-sexp ocamlPackages.yojson ocamlPackages.yuscii ocamlPackages.z3 ocamlPackages.zed ocamlPackages.zmq ocamlPackages.zmq-lwt ocamlPackages.zstandard ocamlformat ocamlformat_0_11_0 ocamlformat_0_12 ocamlformat_0_13_0 ocamlformat_0_14_0 ocamlformat_0_14_1 ocamlformat_0_14_2 ocamlformat_0_14_3 ocamlformat_0_15_0 ocamlformat_0_15_1 ocamlformat_0_16_0 ocamlformat_0_17_0 ocamlformat_0_18_0 ocamlformat_0_19_0 ocamlformat_0_20_0 ocsigen-i18n opa opam-installer orpie patdiff prooftree python310Packages.bap python310Packages.guestfs python39Packages.bap python39Packages.guestfs satysfi sawjap sks soupault unison usync vagrant virt-top wayland-proxy-virtwl why3
</details>

ulrikstrid commented:
And a pretty comprehensive test run that uses this PR as the base for nixpkgs: https://github.com/anmonteiro/nix-overlays/runs/5300761584?check_suite_focus=true
There are some overrides as well but it's largely just using this

ulrikstrid commented:
As a note about size, this decreased the closure size of a static build by a bit more than 100mb (it includes OCaml for some reason) and a non-static build by about 1,4gb (2gb -> 600mb including OCaml).

symphorien commented on pkgs/development/ocaml-modules/lablgtk-extras/1.4.nix:
```diff
@@ -9,7 +9,8 @@ stdenv.mkDerivation rec {
 
-  buildInputs = [ ocaml findlib camlp4 ];
+  nativeBuildInputs = [ ocaml findlib ];
+  buildInputs = [ camlp4 ];
```
camlp4 should go to nativeBuildInputs

symphorien commented on pkgs/development/ocaml-modules/lablgtk-extras/default.nix:
```diff
@@ -16,7 +16,8 @@ stdenv.mkDerivation rec {
 
-  buildInputs = [ ocaml findlib camlp4 ];
+  nativeBuildInputs = [ ocaml findlib ];
+  buildInputs = [ camlp4 ];
```
camlp4 should go to nativeBuildinputs

symphorien commented on pkgs/development/ocaml-modules/lablgtk/2.14.0.nix:
```diff
@@ -16,8 +16,8 @@ stdenv.mkDerivation (rec {
-  nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ ocaml findlib gtk2 libgnomecanvas gtksourceview camlp4 ];
+  nativeBuildInputs = [ pkg-config ocaml findlib ];
+  buildInputs = [ gtk2 libgnomecanvas gtksourceview camlp4 ];
```
camlp4 should go to nativeBuildInputs

symphorien reviewed (commented):
This looks good to me overall. I see that you uncovered a bug where checkinputs don't works correctly with strictDeps, I opened https://github.com/NixOS/nixpkgs/issues/161570 and I have an idea for a fix, but that's for later. I left a few comments, some instances of camlp4 should be moved to nativeBuildInputs (but since camlp4 is restricted to older ocaml releases, you didn't test this)

ulrikstrid commented:
I'll move the instances of camlp4 you found tomorrow (unless we can get @anmonteiro to do it 🙏). I found some instances where we needed it both in nativeBuildInputs and buildInputs to pass the checks btw.

Thank you again for the review @symphorien!

ulrikstrid commented:
@symphorien we seem to have the same issue as with `checkInputs` with `configurationPhase`. In some packages, even if we have OCaml as a nativeBuildDependency we can't run `ocaml ...` in `configurationPhase`.

ulrikstrid commented:
I took the freedom to add `strictDeps = true;` in janePackages and some other places that spread. This let me fix a bunch of extra packages that I had not found previously.

`torch` is still broken on my machine but I just checked master and I have the exact same test failure there. This should be ready for review now.

symphorien reviewed (approved):
nix-review is happy
I'll merge over the weekend unless someone complains.

symphorien commented:
Issue is #161570 and PR is #161575

vbgl commented:
This broke `ocaml-ng.ocamlPackages_4_07.lwt`.

symphorien commented:
fixed in #162370

vbgl commented:
Thanks for the quick fix.
Bisecting the failure of `ocaml-ng.ocamlPackages_4_07.ppx_let` led me here again.

vbgl commented:
This also broke `ocaml-ng.ocamlPackages_4_07.ppx_deriving`.

symphorien commented:
I don't know what makes ppx_let fail, can you have a look @ulrikstrid ?

ulrikstrid commented:
Trying to reproduce the error on with `ocaml-ng.ocamlPackages_4_07.ppx_let` but all I'm seeing is that `ppxlib`  needs dune 2 for version `0.8.1`. Is this what you are seeing as well @vbgl @symphorien ?

Also, I asked in the followup PR if any of you know a good command to build all/most packages across ocaml versions so that I can catch these earlier.

vbgl commented:
This broke apron: it builds fine but is not installed properly:

`nix-shell -p ocaml-ng.ocamlPackages_4_12.findlib ocaml-ng.ocamlPackages_4_12.apron --run 'ocamlfind query -r apron'`
> ocamlfind: Package `apron' not found

ulrikstrid commented:
I can't reproduce this on master
```sh
nix-shell -p ocaml-ng.ocamlPackages_4_12.findlib ocaml-ng.ocamlPackages_4_12.apron --run 'ocamlfind query -r apron'
/nix/store/mps7hcb0qp7cf4l5k582asxwxhdl08kx-ocaml-4.12.0/lib/ocaml
/nix/store/mps7hcb0qp7cf4l5k582asxwxhdl08kx-ocaml-4.12.0/lib/ocaml
/nix/store/hiisfrrljwbj99h4q0spqizzj2xgr7c5-ocaml4.12.0-mlgmpidl-1.2.12/lib/ocaml/4.12.0/site-lib/gmp
/nix/store/lf0cyn31apa7yi2f25w6dih528cjbn3d-ocaml4.12.0-apron-0.9.13-dev/lib/ocaml/4.12.0/site-lib/apron
/nix/store/lf0cyn31apa7yi2f25w6dih528cjbn3d-ocaml4.12.0-apron-0.9.13-dev/lib/ocaml/4.12.0/site-lib/apron
```

vbgl commented:
You may need to set the `NIX_PATH` environment variable.

ulrikstrid commented:
This would be the same thing right?
```
➜  nixpkgs git:(master)  nix-shell -I ./default.nix -p ocaml-ng.ocamlPackages_4_12.findlib ocaml-ng.ocamlPackages_4_12.apron
[nix-shell:~/dev/nixpkgs]$ ocamlfind query -r apron
/nix/store/mps7hcb0qp7cf4l5k582asxwxhdl08kx-ocaml-4.12.0/lib/ocaml
/nix/store/mps7hcb0qp7cf4l5k582asxwxhdl08kx-ocaml-4.12.0/lib/ocaml
/nix/store/hiisfrrljwbj99h4q0spqizzj2xgr7c5-ocaml4.12.0-mlgmpidl-1.2.12/lib/ocaml/4.12.0/site-lib/gmp
/nix/store/lf0cyn31apa7yi2f25w6dih528cjbn3d-ocaml4.12.0-apron-0.9.13-dev/lib/ocaml/4.12.0/site-lib/apron
/nix/store/lf0cyn31apa7yi2f25w6dih528cjbn3d-ocaml4.12.0-apron-0.9.13-dev/lib/ocaml/4.12.0/site-lib/apron
```

vbgl commented:
Did you try: `nix-shell -I nixpkgs=./default.nix …`?

vbgl commented:
Or `-I ..`?

ulrikstrid commented:
`nix-shell -I nixpkgs=./default.nix -p ocaml-ng.ocamlPackages_4_12.findlib ocaml-ng.ocamlPackages_4_12.apron` show the error, let me investigate

ulrikstrid commented:
> Thanks for the quick fix. Bisecting the failure of `ocaml-ng.ocamlPackages_4_07.ppx_let` led me here again.

It almost looks like dune2 was somehow available for ppxlib even in a case where it shouldn't have been. I have a PR coming up fixing this, I just need to test it with a couple of other OCaml versions before submitting.

vbgl commented:
`ocamlPackages.elina` seems broken too.

ulrikstrid commented:
@vbgl `elina` and a bunch of other packages are fixed in https://github.com/NixOS/nixpkgs/pull/162745, would love a review when you have time

Et7f3 commented on pkgs/development/ocaml-modules/ppx_tools/default.nix:
```diff
@@ -5,6 +5,7 @@ let param =
     sha256 = "15v7yfv6gyp8lzlgwi9garz10wpg34dk4072jdv19n6v20zfg7n1";
     useDune2 = true;
+    nativeBuildInputs = [cppo];
     buildInputs = [cppo];
```
Is it a left over ?

ulrikstrid replied:
Can't remember but might be that there's a library component of `cppo` that was also needed?

Et7f3 replied:
Ok I will do another round. I see other packages with the same issue.
